node/node.go

```diff
@@ -1393,6 +1435,63 @@ func createAndStartPrivValidatorSocketClient(
 	return pvscWithRetries, nil
 }
 
+func createIpfsNode(config *cfg.Config, arePluginsAlreadyLoaded bool, logger log.Logger) (*ipfscore.IpfsNode, error) {
+	repoRoot := config.IPFSRepoRoot()
+	logger.Info("creating node in repo", "ipfs-root", repoRoot)
+	if !fsrepo.IsInitialized(repoRoot) {
+		// TODO: sentinel err
+		return nil, fmt.Errorf("ipfs repo root: %v not intitialized", repoRoot)
+	}
+	if !arePluginsAlreadyLoaded {
+		if err := setupPlugins(repoRoot, logger); err != nil {
+			return nil, err
+		}
+	}
+	// Open the repo
+	repo, err := fsrepo.Open(repoRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the node
+	nodeOptions := &ipfscore.BuildCfg{
+		Online: true,
+		// This option sets the node to be a full DHT node (both fetching and storing DHT Records)
+		Routing: libp2p.DHTOption,
+		// This option sets the node to be a client DHT node (only fetching records)
+		// Routing: libp2p.DHTClientOption,
+		Repo: repo,
+	}
+	// TODO: change to context.WithCancel if we want to use the context for lifecycle management
```

Comment:
is this todo needed? Isn't the ipfs node closed when `node.OnStop` is called?

Reply:
I think you are onto something here. There was something dodgy about lifecycle management of ipfs nodes 🤔 
As you probably noticed, this is related to the other TODO here: https://github.com/lazyledger/lazyledger-core/blob/d608cdb03b479a54765b7b1448a6ba7fec3602e8/node/node.go#L1020-L1024 (in OnStop) which itself is related to how ipfs uses contexes to shut down the node: https://github.com/lazyledger/go-ipfs/blob/dd295e45608560d2ada7d7c8a30f1eef3f4019bb/core/builder.go#L48-L57

Looking at that ipfs code again, I've realized, we need to make sure to always create a fresh node instance as the context cancellation will happen only once per node. Which, I just checked is the case:
https://github.com/lazyledger/lazyledger-core/blob/d608cdb03b479a54765b7b1448a6ba7fec3602e8/node/node.go#L955

Actually, you are right. We can remove these todos. Should probably add a comment why the node instance has to be recreated and cannot be reused.

Reply:
Removed the todos and replaced them with some hopefully helpful comments: 4675af8